docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+compute cost.
+
+## Motivation
+Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances (ShootKapi
+hereafter). This need is a consequence of the following concerns:
+1. Gardener supplies K8s clusters of highly diverse compute scale and utilization - ranging from a couple of small
+   practically idle nodes, to hundreds of heavily utilized giants. Furthermore, the ratio of 'cluster compute capacity'
+   to 'ShootKapi load' substantially varies between different applications utilizing Gardener shoots. As a result,
+   compute resources required by ShootKapis of different shoots vary by two orders of magnitude.
+2. The compute cost of ShootKapi workloads constitutes a major part of Gardener's overall compute cost. Small relative
+   waste in sizing ShootKapis results in considerable absolute cost.
+3. Kube-apiserver does not scale well horizontally, in a sense that extra replicas come with substantial compute overhead.   
+
+In short, ShootKapi needs to be scaled:
+- Within a broad range
+- With reasonable accuracy, relative to actual resource consumption
+
+The implication is that ShootKapi requires both horizontal and vertical scaling. The existing implementation utilized 
+to that end is [HVPA].
+
+### Existing Solution
+[HVPA] strives to achieve a fusion of HPA and VPA recommendations:
+`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`
+The underlying goal is to allow HPA and VPA to be driven by the same CPU and memory usage metrics, and overcome the
+HPA-VPA incompatibility known to exist in that mode.
+
+### Deficiencies of the Existing Solution
+HVPA strives to achieve a linear combination of HPA's and VPA's scaling policies:
+`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`.
+However there is a nonobvious obstacle before this intent: both HPA and VPA's policies are practically unavailable.
+The only information available is "how would HPA/VPA change the existing workload right now". So, in place of the
+original intent, HVPA resorts to:
+`dHvpaRecommendedCapacity = k * dHpaRecommendedCapacity + (1 - k) * dVpaRecommendedCapacity` (the 'd' symbols stand for
+"rate of change").
+
+Had HPA and VPA recommendations been always deterministically available, up to date, and with absolute accuracy,
+that would result in the originally intended behavior. Simply put, in a 3-D space where one coordinate is the real-world
+utilization data, and the other two are the horizontal and vertical scale suggested by the recommender, this is a line -
+a line which is a weighted average of the two lines which are individual HPA and VPA recommendations (Fig. 1).
+
+![HVPA scaling](./assets/gep22-hvpa-scaling-space.png "Design outline")
+
+_Fig. 1: HVPA policy as a weighted average of HPA and VPA policies. HPA and VPA policies in blue. HVPA intended policy
```

Comment:
A sketch of the extreme case where only scaling on one axis occurs, looks completely unremarkable - just a single straight line with intermediate points in it. And a more visually engaging scenario - one with some scaling on the other axis, and maybe even paradoxical downscaling (where VPA downscales immediately after HPA scales out), becomes verbose to explain and harder to understand, because it demonstrates multiple points at the same time.

I opted for adding the following explanation instead of a sketch (will be available in next push):
>As example of extreme single axis scaling, consider the following. A workload with constant memory consumption is being scaled by a HVPA configured at 50% HPA + 50% VPA, starting at 1 replica with 1 core. The expectation is that HVPA will maintain the number of replicas approximately equal to the number of cores per replica. Consider the following sequence of events, as CPU demand increases:
    1. CPU usage starts at 0.7 cores. HVPA is stable at 1 replica, 1 core
    2. CPU usage increases to 1.5 cores. HPA acts first, recommending 2 replicas, a change of +1. 50% of +1 is +0.5 replicas, rounded up to +1. HVPA applies +1 replica. The workload is now at consumption=1.5,Replicas=2,CoresPerReplica=1
    3. VPA acts, but does not recommend upscaling because CPU is currently over-provisioned. It does not recommend downscaling because MinAllowed is 1 core.
    4. CPU usage increases to 2.1 cores. HPA acts first, recommending 3 replicas, a change of +1. HVPA applies
       2 + ceil(50% * +1) = 3 replicas. The workload is now at Consumption=2.1,Replicas=3,CoresPerReplica=1
    ...and so on. HVPA acts as 100% HPA + 0% VPA, instead of the configured 50%/50%.

Reply:
Closing conversation to keep the PR manageable. Please reopen if necessary.